src/expr/src/interpret.rs

```diff
@@ -768,6 +776,40 @@ impl<'a> ColumnSpecs<'a> {
             },
         }
     }
+
+    fn set_literal(expr: &mut MirScalarExpr, value: Result<Datum, EvalError>) {
+        match expr {
+            MirScalarExpr::Literal(row, col_type) => {
+                *row = value.map(|d| {
+                    assert!(
+                        d.is_instance_of(col_type),
+                        "{d:?} must be an instance of {col_type:?}"
+                    );
+                    Row::pack_slice(&[d])
+                })
+            }
+            _ => panic!("not a literal"),
+        }
+    }
+
+    fn set_argument(expr: &mut MirScalarExpr, arg: usize, value: Result<Datum, EvalError>) {
```

Comment:
Assuming you call a `set_argument` (which delegates to `set_literal`) on `MirScalarExpr` nodes that have been initialized with a `placeholder()` call, how are these calls going to mutate the `MirScalarExpr`? The `set_literal` will only work if the `Literal` variants holds an `Ok(...)` in its `row` field.

Reply:
> The `set_literal` will only work if the Literal variants holds an Ok(...) in its row field.

I don't think this is right... `set_literal` should work regardless of the result type in the existing literal or in `value`. (I realize I've named the variable `row`, which suggests that it's always an `Ok(Row(...))`... maybe renaming would make that more explicit?)

Reply:
I went ahead and rewrote the `set_literal` logic slightly to make it more obvious that the `Err` case was handled, and added a minor optimization. Hopefully a bit less tricky to reason about this way!